lib/scholar/cross_decomposition/pls_svd.ex

```diff
@@ -0,0 +1,356 @@
+             :x_weights,
+             :y_weights
+           ]}
+  defstruct [
+    :x_mean,
+    :y_mean,
+    :x_std,
+    :y_std,
+    :x_weights,
+    :y_weights
+  ]
+
+  opts_schema = [
+    num_components: [
+      default: 2,
+      type: :pos_integer,
+      doc: "The number of components to keep. Should be in `[1,
+        min(n_samples, n_features, n_targets)]`."
+    ],
+    scale: [
+      default: true,
+      type: :boolean,
+      doc: "Whether to scale `x` and `y`."
+    ]
+  ]
+
+  @opts_schema NimbleOptions.new!(opts_schema)
+
+  @doc """
+    Fit model to data.
+    Takes as arguments: 
+
+    * `x` - training samples, `{num_samples, num_features}` shaped tensor
+    
+    * `y` - targets, `{num_samples, num_targets}` shaped `y` tensor
+
+  ## Options
+
+  #{NimbleOptions.docs(@opts_schema)}
+
+  ## Return Values
+
+    The function returns fitted estimator represented by struct with the following parameters:
+
+    * `:x_mean` - tensor of shape `{num_features}` which represents `x` tensor mean values calculated along axis 0.
+
+    * `:y_mean` - tensor of shape `{num_targets}` which represents `x` tensor mean values calculated along axis 0.
+
+    * `:x_std` - tensor of shape `{num_features}` which represents `x` tensor standard deviation values calculated along axis 0.
+
+    * `:y_std` -  tensor of shape `{num_targets}` which represents `y` tensor standard deviation values calculated along axis 0.
+
+    * `:x_weights` -  tensor of shape `{num_features, num_components}` the left singular vectors of the SVD of the cross-covariance matrix.
+
+    * `:y_weights` -  tensor of shape `{num_targets, num_components}` the right singular vectors of the SVD of the cross-covariance matrix.
+
+  ## Examples
+
+      iex> x = Nx.tensor([[0.0, 0.0, 1.0],
+      ...>                [1.0, 0.0, 0.0],
+      ...>                [2.0, 2.0, 2.0],
+      ...>                [2.0, 5.0, 4.0]])
+      iex> y = Nx.tensor([[0.1, -0.2],
+      ...>                [0.9, 1.1],
+      ...>                [6.2, 5.9],
+      ...>                [11.9, 12.3]])
+      iex> model = Scholar.CrossDecomposition.PLSSVD.fit(x, y)
+      iex> model.x_mean
+      #Nx.Tensor<
+        f32[3]
+        [1.25, 1.75, 1.75]
+      >
+      iex> model.y_std
+      #Nx.Tensor<
+        f32[2]
+        [5.467098712921143, 5.661198616027832]
+      >
+      iex> model.x_weights
+      #Nx.Tensor<
+        f32[3][2]
+        [
+          [0.521888256072998, -0.11256571859121323],
+          [0.6170258522033691, 0.7342619299888611],
+          [0.5889922380447388, -0.6694686412811279]
+        ]
+      >
+  """
+
+  deftransform fit(x, y, opts \\ []) do
+    fit_n(x, y, NimbleOptions.validate!(opts, @opts_schema))
+  end
+
+  defnp fit_n(x, y, opts) do
+    {x, y} = check_x_y(x, y, opts)
+    num_components = opts[:num_components]
+    {x, y, x_mean, y_mean, x_std, y_std} = center_scale_x_y(x, y, opts)
+
+    c =
+      Nx.transpose(x)
+      |> Nx.dot(y)
```

Comment:
```suggestion
    c = Nx.dot(x, [1], y, [1])
```

Reply:
@josevalim I think we could add an in-place rewrite in Nx.Defn.Expr that checks if the immediate child arg is Nx.transpose, and aplies the permutation to the contract and axes for dot/2.

This is a really common pattern that I tend to find in reviews